File: RetrofitExample/app/src/main/java/com/gorkymunoz/retrofit_example/MainActivity.java
package com.gorkymunoz.retrofit_example;

import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.gorkymunoz.retrofit_example.adapters.CharacterAdapter;
import com.gorkymunoz.retrofit_example.network.models.character.Character;
import com.gorkymunoz.retrofit_example.viewmodel.RickMortyViewModel;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    private final String TAG = MainActivity.class.getSimpleName();

    RickMortyViewModel rickAndMortyViewModel;
    CharacterAdapter characterAdapter;
    RecyclerView rvCharacters;
    List<Character> characterList = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        rvCharacters = findViewById(R.id.rv_characters);
        rickAndMortyViewModel = new ViewModelProvider(this).get(RickMortyViewModel.class);
        rickAndMortyViewModel.fetchCharacters();
        rickAndMortyViewModel.getCharacterResponseMutableLiveData().observe(this, characterResponse -> {
            List<Character> charactersUpdated = characterResponse.getCharacters();
            characterList.clear();
            characterList.addAll(charactersUpdated);
            characterAdapter.notifyDataSetChanged();
        });

        initializeRV();
    }

    void initializeRV() {
        if (characterAdapter == null) {
            characterAdapter = new CharacterAdapter(this, characterList);
            rvCharacters.setLayoutManager(new GridLayoutManager(this, 3, LinearLayoutManager.VERTICAL, false));
            rvCharacters.setAdapter(characterAdapter);
            rvCharacters.setItemAnimator(new DefaultItemAnimator());
        } else {
            characterAdapter.notifyDataSetChanged();
        }
    }
}

File: RetrofitExample/app-kotlin/src/main/java/com/gorkymunoz/app_kotlin/presentation/main/MainActivity.kt
package com.gorkymunoz.app_kotlin.presentation.main

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import com.gorkymunoz.app_kotlin.data.repositories.CharacterRepository
import com.gorkymunoz.app_kotlin.data.usecase.GetAllCharactersUseCase
import com.gorkymunoz.app_kotlin.databinding.ActivityMainBinding
import com.gorkymunoz.app_kotlin.network.RetrofitClient
import com.gorkymunoz.app_kotlin.network.model.character.CharacterList
import com.gorkymunoz.app_kotlin.presentation.common.ResultUI
import com.gorkymunoz.app_kotlin.presentation.common.getViewModel

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var viewModel: MainViewModel
    private val characterAdapter by lazy { CharacterAdapter() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        viewModel = getViewModel {
            MainViewModel(
                GetAllCharactersUseCase(
                    CharacterRepository(RetrofitClient.createService())
                )
            )
        }.also {
            it.getCharacterList()
        }

        binding.rvCharacter.apply {
            layoutManager =
                GridLayoutManager(this@MainActivity, 3, LinearLayoutManager.VERTICAL, false)
            adapter = characterAdapter
        }

        viewModel.characters.observe(this, Observer(::updateUI))
    }

    private fun updateUI(result: ResultUI<CharacterList>) {

        when (result) {
            is ResultUI.Loading -> print("LOADING")
            is ResultUI.Error -> Log.e("ERROR", result.exception.message, result.exception)
            is ResultUI.Success -> updateCharacters(result.data)
        }
    }

    private fun updateCharacters(data: CharacterList) {
        characterAdapter.submitList(data.listOfCharacters)
    }

}